### PR TITLE
Update to make use of display-url-api plugin

### PR DESCRIPTION
**Pull request by i386**
This will make all the instant messaging plugins compatible with Blue ocean.

Upgrading the toolchain ment we had to fix a number of FindBugs issues.

Depends on
- https://issues.jenkins-ci.org/browse/HOSTING-179
- https://github.com/jenkinsci/instant-messaging-parent-plugin/pull/2

@jenkinsci/code-reviewers

### CODE REVIEW

**Comment by kutzi** [comment]
So I need to wait for https://issues.jenkins-ci.org/browse/HOSTING-179 before I can merge this?

**Comment by batmat** [comment]
@kutzi Yeah, seems indeed more sensible.

**Comment by i386** [comment]
@kutzi @batmat we need a release of the parent pom to `1.28` and the build here will start passing. When it passes, it is ready for merge.

File: src/main/java/hudson/plugins/im/IMConnectionProvider.java [hunk 1]
```diff
@@ -174,11 +174,12 @@ public void run() {
...
-                            // exponentially increase timeout, but longer than 16 minutes
-                            if (timeout < 15) {
-                            	timeout *= 2;
-                            }
+                            if (!this.semaphore.tryAcquire(timeout * 60, TimeUnit.SECONDS)) {
+								// exponentially increase timeout, but longer than 16 minutes
+								if (timeout < 15) {
+									timeout *= 2;
+								}
+							}
```
Comment (batmat) [inline src/main/java/hudson/plugins/im/IMConnectionProvider.java]: IIUC this is only WS changes?

File: src/main/java/hudson/plugins/im/IMConnectionProvider.java [hunk 2]
```diff
@@ -174,11 +174,12 @@ public void run() {
                             this.semaphore.drainPermits();
                             
                             // wait up to timeout time OR until semaphore is released again (happens e.g. if global config was changed)
-                            this.semaphore.tryAcquire(timeout * 60, TimeUnit.SECONDS);
-                            // exponentially increase timeout, but longer than 16 minutes
-                            if (timeout < 15) {
-                            	timeout *= 2;
-                            }
+                            if (!this.semaphore.tryAcquire(timeout * 60, TimeUnit.SECONDS)) {
+								// exponentially increase timeout, but longer than 16 minutes
```
Comment (batmat) [inline src/main/java/hudson/plugins/im/IMConnectionProvider.java]: Unrelated to James' change, I guess "no" is missing before "longer"?

File: src/main/java/hudson/plugins/im/bot/HealthCommand.java [hunk 3]
```diff
@@ -26,17 +26,22 @@
...
         StringBuilder msg = new StringBuilder(32);
+        AbstractBuild<?, ?> lastBuild = project.getLastBuild();
         msg.append(project.getFullDisplayName());
         if (project.isDisabled()) {
             msg.append("(disabled)");
         } else if (project.isBuilding()) {
-            msg.append("(BUILDING: ").append(project.getLastBuild().getDurationString()).append(")");
+            msg.append("(BUILDING");
+            if (lastBuild != null) {
+                msg.append(": ");
```
Comment (batmat) [inline src/main/java/hudson/plugins/im/bot/HealthCommand.java]: Aren't mixing things here? Though this change does seem to make sense, is it related to supporting display-url-plugin?

Comment (batmat) [inline src/main/java/hudson/plugins/im/bot/HealthCommand.java]: OK, re-reading the commit message, @i386 says he had to fix some FB issues so I guess it's one of them. 
NIT: maybe to PR would have been clearer, one to upgrade and fix those, one only for display-url-plugin changes.

File: src/main/java/hudson/plugins/im/bot/SetAliasCommand.java [hunk 4]
```diff
@@ -89,7 +89,7 @@ public String getHelp() {
         public AliasCommand(BotCommand cmd, String commandName, String[] arguments) {
 			this.command = cmd;
 			this.commandName = commandName;
-			this.arguments = arguments;
+			this.arguments = Arrays.copyOf(arguments, arguments.length);
```
Comment (batmat) [inline src/main/java/hudson/plugins/im/bot/SetAliasCommand.java]: Semantic change. I guess this won't hurt, but hope the current code doesn't have a corner case relying on that (OK, that wouldn't seem great).

Comment (i386) [inline src/main/java/hudson/plugins/im/bot/SetAliasCommand.java]: Exposure of a mutable array caused a find bugs warning here (many other changes in this PR relate to find bugs)

**Comment by batmat** [comment]
:+1: 

though as I wrote I guess two PRs, one for upgrading pom + supporting display-url-plugin might have clarified.

**Comment by i386** [comment]
I can split them if you'd like

File: src/main/java/hudson/plugins/im/bot/SetAliasCommand.java (hunk 4 above)
Comment (batmat) [inline src/main/java/hudson/plugins/im/bot/SetAliasCommand.java]: Yeah, this is what I understood afterwards (I wrote it in my previous line comment above). FB related.

**Comment by batmat** [comment]
@i386 I'm not the maintainer, so in the end it's @kutzi to say :). Maybe next time? ;-)

**Comment by i386** [comment]
@kutzi I'd be happy to split this up if you'd like

File: pom.xml [hunk 5]
```diff
@@ -47,6 +47,11 @@
     </build>
     
     <dependencies>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>display-url-api</artifactId>
+            <version>0.2</version>
```
Comment (kutzi) [inline pom.xml]: Shouldn't that be an optional dependency?
Otherwise the plugin won't run without the display-url-api plugin, right?

**Comment by kutzi** [comment]
Yeah, better if you could split this up, if there are multiple logical changes in this

File: src/test/java/hudson/plugins/im/tools/MessageHelperTest.java [hunk 6]
```diff
@@ -70,19 +85,27 @@ public void testConcat() {
 	
 	@SuppressWarnings({ "rawtypes", "unchecked" })
     @Test
-	public void testUrlShouldBeUrlEncoded() {
+	public void testUrl() throws Exception {
```
Comment (kutzi) [inline src/test/java/hudson/plugins/im/tools/MessageHelperTest.java]: What does this test actually? 'testUrl' is not really descriptive

Comment (kutzi) [inline src/test/java/hudson/plugins/im/tools/MessageHelperTest.java]: What does this test actually? 'testUrl' is not a really descriptive test name!

**Comment by jglick** [comment]
FTR, closing due to https://github.com/jenkinsci/display-url-api-plugin/pull/13.

**Comment by ernestask** [comment]
I don’t understand the rationale, half of the change here is still applicable.

**Pull request closed without merging**